src/rdkafka_telemetry.c

```diff
@@ -192,32 +249,164 @@ void rd_kafka_handle_push_telemetry(rd_kafka_t *rk, rd_kafka_resp_err_t err) {
         }
 }
 
+/**
+ * @brief This method starts the termination for telemetry and awaits
+ * completion.
+ *
+ * @locks none
+ * @locks_acquired rk_telemetry.lock
+ * @locality app thread (normal case) or the main thread (when terminated
+ *           during creation).
+ */
+void rd_kafka_telemetry_await_termination(rd_kafka_t *rk) {
+        rd_kafka_op_t *rko;
+
+        /* In the case where we have a termination during creation, we can't
+         * send any telemetry. */
+        if (thrd_is_current(rk->rk_thread)) {
+                /* We can change state since we're on the main thread. */
+                rk->rk_telemetry.state = RD_KAFKA_TELEMETRY_TERMINATED;
+                return;
+        }
+
+        rko         = rd_kafka_op_new(RD_KAFKA_OP_TERMINATE_TELEMETRY);
+        rko->rko_rk = rk;
+        rd_kafka_q_enq(rk->rk_ops, rko);
+
+        /* Await termination sequence completion. */
+        rd_kafka_dbg(rk, TELEMETRY, "TELTERM",
+                     "Awaiting termination of telemetry.");
+        mtx_lock(&rk->rk_telemetry.lock);
+        cnd_wait(&rk->rk_telemetry.termination_cnd, &rk->rk_telemetry.lock);
```

Comment:
Will it make sense to do a timed_wait if for some case RD_KAFKA_TELEMETRY_TERMINATING_PUSH_SCHEDULED gets blocked somewhere?

Reply:
Agreed

How much should we wait for? 1s?

Reply:
Hmm not sure, but I think 1s should be good. We can probably tweak it later if the terminating push takes more time.

Reply:
Done, 1s